src/node_sea.cc

```diff
@@ -78,6 +95,172 @@ std::tuple<int, char**> FixupArgsForSEA(int argc, char** argv) {
   return {argc, argv};
 }
 
+class PrinterTryCatch : public TryCatch {
+ public:
+  explicit PrinterTryCatch(Isolate* isolate)
+      : TryCatch(isolate), isolate_(isolate) {}
+  ~PrinterTryCatch() {
+    if (!HasCaught()) {
+      return;
+    }
+    PrintCaughtException(isolate_, isolate_->GetCurrentContext(), *this);
+  }
+
+ private:
+  Isolate* isolate_;
+};
+
+template <typename T, void (*function)(T*)>
+class PointerWithDeleter {
+ public:
+  explicit PointerWithDeleter(T* ptr) : ptr_(ptr) {}
+  T* get() const { return ptr_; }
+  ~PointerWithDeleter() { function(ptr_); }
+
+ private:
+  T* ptr_;
+};
+
+// We might want a lighter-weight JSON parser because we only looking into
+// a few top-level fields of a simple JSON here. But for now using V8
+// is good enough for reading SEA configs.
+class JSONParser {
```

Comment:
I know this isn't ready for review yet but just a high level question on the overall approach for the serializer - wouldn't it be simpler if we did the config parsing logic in JavaScript in a script that can be added to [`lib/internal/main`](https://github.com/nodejs/node/tree/main/lib/internal/main)?

Reply:
In the case of snapshot building we need to at least run the `main/mksnapshot.js` so we'd end up running two main scripts in the same process, which I am not sure is safe to do at this point.

Reply:
> which I am not sure is safe to do at this point

Why could it be unsafe? If it's unsafe because some things haven't been initialized in the current instance of node, could this be done by spawning a child node instance with `--build-snapshot`?

Reply:
And extra child process would make it harder to debug though, I am not sure if it's worth the complexity for just...parsing some JSON.

Reply:
> And the environment setup necessary for spawning a child process would not be simpler than the minimal V8 setup code here.

I haven't run this locally yet but wouldn't doing that be pretty much the same as having a file like this?
**`lib/internal/main/prepare-sea-blob.js`**
```js
const {
  prepareMainThreadExecution,
  markBootstrapComplete,
} = require('internal/process/pre_execution');

const { execFile } = require('child_process');

prepareMainThreadExecution(false, true);
markBootstrapComplete();

...

execFile(
  process.argv,
  [ '--snapshot-blob',
    snapshotBlob,
    '--build-snapshot',
    entryPointScript ],
  (error, stdout, stderr) => {
  ...
});

...
```

To be on the same page, are you saying that this is not gonna be simpler than the minimal V8 setup code on this PR?

Reply:
That would be the "harder to debug" part (usually I'd try to avoid spawning a child process if I can). The reasonable compromise would be to just run the snapshot building in the main script for preparing a SEA blob, but that also requires some refactoring that seems better for a follow-up. Also I usually prefer just not to spin up any Node.js environment if the task is simple, because spinning up a Node.js environment involves too many things - superficially it seems simple if we just add a `StartExecution`, but in reality that does a lot of unnecessary things underneath if we do not technically need a Node.js environment for what we want to do, and parsing JSON -> copying bytes don't really need a Node.js environment, no?

Reply:
> The reasonable compromise would be to just run the snapshot building in the main script for preparing a SEA blob, but that also requires some refactoring that seems better for a follow-up.

That idea sounds good to me.

> Also I usually prefer just not to spin up any Node.js environment if the task is simple, because spinning up a Node.js environment involves too many things - superficially it seems simple if we just add a `StartExecution`, but in reality that does a lot of unnecessary things underneath if we do not technically need a Node.js environment for what we want to do, and parsing JSON -> copying bytes don't really need a Node.js environment, no?

In this case, since high performance isn't really a major requirement I thought readability should have a higher priority. Since parsing a JSON doesn't seem to be an SEA-specific thing, another solution is to move the JSON parser to some other file like `src/util.h` (or maybe a new file like `src/json_parser.h`), wdyt?